Add a resample preprocessor to get the LFP or similar

maurotoro opened the pull request:
Hello peeps,

I needed to get LFPs and other freq samples from some recordings, but this preprocessor was missing. 

I just ported the previous API to the current, and added a test, that should work.

As always, many thanks for the amazing work!

Mucha Suerte!

alejoe91 commented on spikeinterface/toolkit/preprocessing/resample.py:
```diff
@@ -0,0 +1,125 @@
+)
+from .filter import fix_dtype
+
+try:
+    from scipy import special, signal
+
+    HAVE_RR = True
+except ImportError:
+    HAVE_RR = False
```
you can safely remove the try except since the `toolkit` module depends already on `scipy`

maurotoro replied:
Did them on the new pull.

alejoe91 commented on spikeinterface/toolkit/preprocessing/resample.py:
```diff
@@ -0,0 +1,125 @@
+    """
+
+    name = "resample"
+    installed = HAVE_RR  # check at class level if installed or not
+    installation_mesg = (
+        "To use the ResampleRecording, install scipy: \n\n pip install scipy\n\n"  # err
+    )
```
```suggestion
```

alejoe91 commented on spikeinterface/toolkit/preprocessing/resample.py:
```diff
@@ -0,0 +1,125 @@
+        resample_rate,
+        dtype=None,
+    ):
+        # Check that scipy is present
+        assert self.installed, self.installation_mesg
```
```suggestion
```

alejoe91 commented on spikeinterface/toolkit/preprocessing/resample.py:
```diff
@@ -0,0 +1,125 @@
+        dtype = fix_dtype(recording, dtype)
+        self.dtype = dtype
+
+        BasePreprocessor.__init__(
+            self, recording, sampling_frequency=resample_rate, dtype=dtype
+        )
```
We have to be careful here because some segments mugh have a `time_vector` (see `BaseRecording.set_times()`). In that case we could either:
- resample the time vector as well
- throw a warning and discard the time vector

maurotoro replied:
I just discarded it.... but it works.

alejoe91 commented on spikeinterface/toolkit/preprocessing/resample.py:
```diff
@@ -0,0 +1,125 @@
+        parent_traces = self.parent_recording_segment.get_traces(
+            parent_start_frame, parent_end_frame, slice(None)
+        )
+        resampled_traces = signal.resample(
+            parent_traces, int(end_frame - start_frame), axis=0
+        )
```
it doesn't seem to use `decimate` if rates are multiples

alejoe91 reviewed (changes requested):
Thank you so much @maurotoro 

I added a few comments!

alejoe91 commented on spikeinterface/toolkit/preprocessing/tests/test_resample.py:
```diff
@@ -0,0 +1,38 @@
+from spikeinterface.core.testing_tools import generate_recording
+
+# from spikeinterface.toolkit.preprocessing import resample
```
```suggestion
import pytest
```

alejoe91 replied:
```suggestion
import pytest

from spikeinterface.toolkit.preprocessing import resample
```

samuelgarcia commented:
Hi Mauricio,
thansk for this. i will have a look soon.

A big challenge fro resampling on the fly it how to deal with chunks and margin.

I think it is very important to add a test that run the resampling with several chunk size.
And then check that the output is the same (or almost the same).
spikeinetrface with this lazy `get_trace()` naturally have chunk processing in mind.
resamping by chunk is hard to design and  theoritically don't give the same results as resampling the entire at once.

Lets discuss this.

maurotoro reviewed (commented):
I'll add the requests and repull in a sec!

maurotoro commented:
@alejoe91  I added your comments, relevant changes were:
  - Adding the signal.decimate, seems like you also had issues with that before on the previous API, I added a trick such that if there's a nan in the result, use signal.resample.
  - Corrected the docstring.
  - Changed the imports.
  - For now it just removes the time_vector for any recording segment.
  
 @samuelgarcia  this is in fact a cool problem, for now I'm not aware of how to solve, but given that with different durations, and lengths the resulting durations and frame_numbers are looking good, it could be working appropiately (yeahright). If you have any ideas on how to show that in did it does work, please give me a comment!
 
 Many thanks for everything!
 Mucha Suerte!

maurotoro commented:
(Seems like I didn't understood the chunk_size issue, sorry...)

maurotoro commented:
Sorry for the long delay, would love to get in touch, my schedule is a bit disorganized and will be likely to jitter a lot these next weeks. But, if we manage to address this through email, I think it would be easier for me to help to work this out.

In any case, maybe a short meeting to get a sense of these issues could be more helpful, how can I give you peeps my email without making it exposed to some internet scammers? 

Mucha Suerte!

alejoe91 commented:
Can you send me an email on the address listed here: https://spikeinterface.readthedocs.io/en/latest/authors.html ?

maurotoro commented:
Took long enough to make this work...

The new resample method and some tests to show it works. Getting traces with margins did help, the tests compare a sinusoidal signal with known frequencies and amplitudes, that they always end up with the correct number of samples and time vectors. Also, there's a comment that plots the signals on top of each other orderly, it shows that they all look similar and well-behaved. The other test compares the RMS with different chunk-sizes for different resampling rates, as in the phase_shift tests, the only relevant detail is that one need the chunk_size to be at least one sec of the new signal, else it breaks.

maurotoro commented:
Sorry, forgot to remove a printing statement and in fact run the tests. Now it should work.

maurotoro reviewed (commented):
this all where alreaduy taken into account later, I didn't notice that I missed the submit.

samuelgarcia commented:
Hi Mauricio,
thanks for this update.
I will have a look on it soon. I promise!

samuelgarcia commented:
Hi Mauricio.
I have read your implementation only now.
I have several changes to ask and also some comments.
Maybe a short call would be more efficient, no ?
Also I change push change directly in your code in you accept.

maurotoro commented:
We can talk tomorrow, time TBA? I will only  be available by the PT evening, you tell me when. Suerte!

maurotoro commented:
closed to make a new one from a branch.
